nexus/src/app/update/mgs_clients.rs

```diff
@@ -0,0 +1,244 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Module providing support for handling failover between multiple MGS clients
+
+use futures::Future;
+use gateway_client::types::SpType;
+use gateway_client::types::SpUpdateStatus;
+use gateway_client::Client;
+use slog::Logger;
+use std::collections::VecDeque;
+use std::sync::Arc;
+use uuid::Uuid;
+
+pub(super) type GatewayClientError =
+    gateway_client::Error<gateway_client::types::Error>;
+
+pub(super) enum PollUpdateStatus {
+    Preparing { progress: Option<f64> },
+    InProgress { progress: Option<f64> },
+    Complete,
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum UpdateStatusError {
+    #[error("different update is now preparing ({0})")]
+    DifferentUpdatePreparing(Uuid),
+    #[error("different update is now in progress ({0})")]
+    DifferentUpdateInProgress(Uuid),
+    #[error("different update is now complete ({0})")]
+    DifferentUpdateComplete(Uuid),
+    #[error("different update is now aborted ({0})")]
+    DifferentUpdateAborted(Uuid),
+    #[error("different update failed ({0})")]
+    DifferentUpdateFailed(Uuid),
+    #[error("update status lost (did the SP reset?)")]
+    UpdateStatusLost,
+    #[error("update was aborted")]
+    UpdateAborted,
+    #[error("update failed (error code {0})")]
+    UpdateFailedWithCode(u32),
+    #[error("update failed (error message {0})")]
+    UpdateFailedWithMessage(String),
+}
+
+#[derive(Debug, thiserror::Error)]
+pub(super) enum PollUpdateStatusError {
+    #[error(transparent)]
+    StatusError(#[from] UpdateStatusError),
+    #[error(transparent)]
+    ClientError(#[from] GatewayClientError),
+}
+
+#[derive(Debug, Clone)]
+pub struct MgsClients {
+    clients: VecDeque<Arc<Client>>,
+}
+
+impl MgsClients {
+    /// Create a new `MgsClients` with the given `clients`.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `clients` is empty.
+    pub fn new<T: Into<VecDeque<Arc<Client>>>>(clients: T) -> Self {
+        let clients = clients.into();
+        assert!(!clients.is_empty());
+        Self { clients }
+    }
+
+    /// Create a new `MgsClients` with the given `clients`.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `clients` is empty.
+    pub fn from_clients<I: IntoIterator<Item = Client>>(clients: I) -> Self {
+        let clients = clients
+            .into_iter()
+            .map(Arc::new)
+            .collect::<VecDeque<Arc<Client>>>();
+        Self::new(clients)
+    }
```

Comment:
This can't accept `&[Client]`, because iterating that only gives `&Client`s, and we need an owned `Client` to pass to `Arc::new`. To make the existing tests still compile, we'd need to take `[Client; N]` and still have to be generic over `N`.

I think/hope it's probably not worth worrying about monomorphization of small helper functions like this - feels like a very tiny drop in a very large bucket compared to diesel / dropshot / serde etc. Maybe we're suffering from a thousand tiny cuts, but I don't really want to compromise on API ergonomics unless we're really sure we need to.

Reply:
Maybe `Vec<Client>`?  But I'm totally fine with not worrying about it here.